File: backend/app/schemas/emby.py
"""Emby 媒体库分析 Schemas。"""

from datetime import datetime

from pydantic import BaseModel, Field


class EmbySeriesStatus(BaseModel):
    """单部剧集在 Emby 中的存在状态。"""

    tmdb_id: int
    title: str | None = None
    year: int | None = None
    poster_url: str | None = None
    in_emby: bool = False
    emby_id: str | None = None
    emby_path: str | None = None
    present_episodes: int | None = None
    missing_eps_count: int = 0
    overview: str | None = None
    tmdb_aired_eps: int | None = None
    tmdb_total_eps: int | None = None
    adjusted_missing: int | None = None
    emby_image_url: str | None = None
    is_subscribed: bool = False
    is_blacklisted: bool = False
    emby_name: str | None = None
    emby_year: int | None = None
    emby_episode_count: int | None = None


class EmbyLibraryAnalysis(BaseModel):
    """Emby 媒体库完整性分析结果。"""

    total_series: int = 0
    subscribed_count: int = 0
    blacklisted_count: int = 0
    series: list[EmbySeriesStatus] = Field(default_factory=list)


class EmbyBlacklistEntry(BaseModel):
    """黑名单条目。"""

    tmdb_id: int
    created_at: datetime


class BlacklistCreate(BaseModel):
    """添加黑名单请求。"""

    tmdb_id: int


class BlacklistActionResponse(BaseModel):
    """黑名单操作响应。"""

    success: bool = True
    message: str = ""


class EmbySubscribeRequest(BaseModel):
    """从 Emby 缺集分析发起订阅的请求。"""

    tmdb_id: int
    title: str
    media_type: str = "tv"
    poster_url: str | None = None
    year: int | None = None


class EmbyActionRequest(BaseModel):
    """Emby 通用操作请求（仅需 tmdb_id）。"""

    tmdb_id: int


class EmbyActionResponse(BaseModel):
    """Emby 操作结果响应。"""

    success: bool = True
    message: str = ""

File: frontend/src/types/emby.ts
/**
 * Emby 媒体库相关类型 — 镜像 backend/app/schemas/emby.py、emby_cache.py。
 */

/** Emby 缓存剧集（含 adjusted_missing 计算字段） */
export interface EmbyCacheResponse {
  tmdb_id: number
  emby_series_name: string | null
  emby_year: number | null
  emby_episode_count: number | null
  emby_image_url: string | null
  emby_path: string | null
  emby_library_name: string | null
  overview: string | null
  poster_url: string | null
  tmdb_total_eps: number | null
  tmdb_aired_eps: number | null
  tmdb_next_air_date: string | null
  is_subscribed: boolean
  is_blacklisted: boolean
  adjusted_missing: number | null
}

/** 缺集分析结果（分页） */
export interface EmbyMissingAnalysis {
  total_series: number
  subscribed_count: number
  missing_count: number
  series: EmbyCacheResponse[]
  page: number
  page_size: number
  total_pages: number
  /** 所有 Emby 媒体库名称列表 */
  libraries: string[]
}

/** TMDB 404 条目 — EmbyCache 有但 TmdbCache 无对应记录 */
export interface Tmdb404Item {
  tmdb_id: number
  emby_series_name: string | null
  emby_year: number | null
  emby_path: string | null
  cd2_path: string | null
  emby_image_url: string | null
  detected_at: string | null
}

/** Emby 同步操作结果 */
export interface EmbySyncResult {
  success: boolean
  total_synced: number
  message: string
}

/** 全量扫描进度状态 */
export interface EmbyScanStatus {
  running: boolean
  progress: number
  current_step: number
  total_steps: number
  step_name: string
  total_items: number
  current_item: number
  started_at: string | null
  error: string | null
}

/** TMDB 404 路径确认结果 */
export interface Tmdb404ResolveResult {
  verified: boolean
  cd2_path: string
  name?: string
  file_type?: 'directory' | 'file'
  sub_items?: number
  sub_file_names?: string[]
  note?: string
}

/** 黑名单条目 */
export interface EmbyBlacklistEntry {
  tmdb_id: number
  created_at: string
}
